src/test.cpp

```diff
@@ -43,31 +42,57 @@ int main() { // Test entry point
     return _global_tests.run();
 }
 
+#define STREAM stderr
+#ifdef CI
+#define PRE ""
+#define POST "\n"
+#else
+#define PRE CLEAR "\r"
+#define POST "\r"
+#endif
+
 unsigned int TestSuite::run() {
-    std::printf("Running %d tests\n", num_tests);
+    //TODO(gu) command line flag
+    FILE *report;
+#ifdef REPORT
+    report = std::fopen("report.txt", "w");  // TODO(gu) absolute path
+    if (!report)
+        std::fprintf(STREAM, "Unable to open report\n");
+#else
+    report = nullptr;
+#endif
+
+    std::fprintf(STREAM, "Running %d tests\n", num_tests);
     unsigned int succeeded = 0;
 
     for (unsigned int i = 0; i < num_tests; i++) {
         GameState state = {};
         _test_gs = &state;
-        std::printf(CLEAR "\r%d/%d:  " YELLOW "testing" RESET " %s\r",
-                i+1, num_tests, tests[i].name);
+        std::fprintf(STREAM, PRE "%2d: %s" POST, i+1, tests[i].name);
+        LOG_TESTS("%2d: %s\n", i+1, tests[i].name);
         bool success = false;
         try {
-            success = tests[i].func(&state);
+            success = tests[i].func(&state, report);
         } catch (const std::runtime_error &ex) { /* Empty */ }
         if (success) {
             succeeded++;
+            LOG_TESTS("test '%s' success\n\n", tests[i].name);
         } else {
-            std::printf(CLEAR "\r" RED "f" RESET " %s (%s @ %d)\n",
-                    tests[i].name, tests[i].file, tests[i].line);
+            std::fprintf(STREAM, PRE BOLDRED "| %s" RESET " failed (%s @ %d)\n",
+                         tests[i].name, tests[i].file, tests[i].line);
+            LOG_TESTS("test '%s' failed\n\n", tests[i].name);
         }
     }
-    std::fprintf(stderr, CLEAR);
-    std::fprintf(stderr, GREEN "Passed:" RESET " %d\n", succeeded);
+    std::fprintf(STREAM, PRE);
+
+    std::fprintf(STREAM, GREEN "Passed: " RESET "%d\n", succeeded);
     if (succeeded != num_tests)
-        std::fprintf(stderr, RED "Failed:" RESET " %d\n", num_tests - succeeded);
+        std::fprintf(STREAM, RED "Failed: " RESET "%d\n", num_tests - succeeded);
+
+    LOG_TESTS("Tests done: %d/%d passed\n", succeeded, num_tests);
```

Comment:
Do you really mean to have double newlines?

Reply:
I do not